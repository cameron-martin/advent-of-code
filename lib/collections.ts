export function maxBy<T>(array: T[], by: (elem: T) => number): T {
    return array.reduce((currentHighest, elem) => by(elem) > by(currentHighest) ? elem : currentHighest);
}

export function minBy<T>(array: T[], by: (elem: T) => number): T {
    return array.reduce((currentLowest, elem) => by(elem) < by(currentLowest) ? elem : currentLowest);
}

export function sum(numbers: number[]): number {
    return numbers.reduce((a, b) => a + b, 0);
}

export function product(numbers: number[]): number {
    return numbers.reduce((a, b) => a * b, 1);
}

type Arrays<T> = {
    [K in keyof T]: Array<T[K]>;
};

export function cartesianProduct<T extends any[]>(...args:  Arrays<T>): T[] {
    if(args.length === 0) return [[]] as any;

    const [init, ...rest] = args;

    const restProduct = cartesianProduct(...rest);

    return init.flatMap(val => restProduct.map(rest => [val, ...rest]) as any);
}

export function keyBy<T>(array: T[], f: (x: T) => string): Record<string, T> {
    const result: Record<string, T> = {};

    array.forEach((item) => {
        result[f(item)] = item;
    });

    return result;
}
