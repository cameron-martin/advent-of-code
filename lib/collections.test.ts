import jsc from 'jsverify';
import { cartesianProduct, product } from "./collections";

it('takes cartesian product', () => {
    expect(cartesianProduct([1, 2, 3], ['foo', 'bar'], [4, 5, 6])).toMatchSnapshot();
});

it('has correct length', () => jsc.assertForall(jsc.array(jsc.array(jsc.number)), args => {
    expect(cartesianProduct(...args).length).toBe(product(args.map(arg => arg.length)));
    
    return true;
}));
