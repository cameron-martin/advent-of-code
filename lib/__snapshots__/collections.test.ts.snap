// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`takes cartesian product 1`] = `
Array [
  Array [
    1,
    "foo",
    4,
  ],
  Array [
    1,
    "foo",
    5,
  ],
  Array [
    1,
    "foo",
    6,
  ],
  Array [
    1,
    "bar",
    4,
  ],
  Array [
    1,
    "bar",
    5,
  ],
  Array [
    1,
    "bar",
    6,
  ],
  Array [
    2,
    "foo",
    4,
  ],
  Array [
    2,
    "foo",
    5,
  ],
  Array [
    2,
    "foo",
    6,
  ],
  Array [
    2,
    "bar",
    4,
  ],
  Array [
    2,
    "bar",
    5,
  ],
  Array [
    2,
    "bar",
    6,
  ],
  Array [
    3,
    "foo",
    4,
  ],
  Array [
    3,
    "foo",
    5,
  ],
  Array [
    3,
    "foo",
    6,
  ],
  Array [
    3,
    "bar",
    4,
  ],
  Array [
    3,
    "bar",
    5,
  ],
  Array [
    3,
    "bar",
    6,
  ],
]
`;
